### ISSUE TITLE
Cannot AddInfluence until previous influence is removed with RemoveInfluence

### ISSUE THREAD

**Issue by anvilvapre** [OpenRA/OpenRA#21429]
ra: Red Alert (release-20231010) 

ww3 map, Valley of the snakes, likely a map issue

```
Exception of type `System.InvalidOperationException`: Cannot AddInfluence until previous influence is removed with RemoveInfluence
   at OpenRA.Mods.Common.Traits.Aircraft.AddInfluence(ValueTuple`2[] landingCells) in /home/runner/work/OpenRA/OpenRA/OpenRA.Mods.Common/Traits/Air/Aircraft.cs:line 870
   at OpenRA.Mods.Common.Traits.Aircraft.AddInfluence(CPos landingCell) in /home/runner/work/OpenRA/OpenRA/OpenRA.Mods.Common/Traits/Air/Aircraft.cs:line 881
   at OpenRA.Mods.Common.Activities.Land.Tick(Actor self) in /home/runner/work/OpenRA/OpenRA/OpenRA.Mods.Common/Activities/Air/Land.cs:line 232
   at OpenRA.Activities.Activity.TickOuter(Actor self) in /home/runner/work/OpenRA/OpenRA/OpenRA.Game/Activities/Activity.cs:line 108
   at OpenRA.Traits.ActivityUtils.RunActivity(Actor self, Activity act) in /home/runner/work/OpenRA/OpenRA/OpenRA.Game/Traits/ActivityUtils.cs:line 29
   at OpenRA.Activities.Activity.TickChild(Actor self) in /home/runner/work/OpenRA/OpenRA/OpenRA.Game/Activities/Activity.cs:line 144
   at OpenRA.Activities.Activity.TickOuter(Actor self) in /home/runner/work/OpenRA/OpenRA/OpenRA.Game/Activities/Activity.cs:line 108
   at OpenRA.Traits.ActivityUtils.RunActivity(Actor self, Activity act) in /home/runner/work/OpenRA/OpenRA/OpenRA.Game/Traits/ActivityUtils.cs:line 29
   at OpenRA.Actor.Tick() in /home/runner/work/OpenRA/OpenRA/OpenRA.Game/Actor.cs:line 262
   at OpenRA.World.Tick() in /home/runner/work/OpenRA/OpenRA/OpenRA.Game/World.cs:line 446
   at OpenRA.Game.InnerLogicTick(OrderManager orderManager) in /home/runner/work/OpenRA/OpenRA/OpenRA.Game/Game.cs:line 631
   at OpenRA.Game.LogicTick() in /home/runner/work/OpenRA/OpenRA/OpenRA.Game/Game.cs:line 646
   at OpenRA.Game.Loop() in /home/runner/work/OpenRA/OpenRA/OpenRA.Game/Game.cs:line 815
   at OpenRA.Game.Run() in /home/runner/work/OpenRA/OpenRA/OpenRA.Game/Game.cs:line 868
   at OpenRA.Game.InitializeAndRun(String[] args) in /home/runner/work/OpenRA/OpenRA/OpenRA.Game/Game.cs:line 306
   at OpenRA.Launcher.Program.Main(String[] args) in /home/runner/work/OpenRA/OpenRA/OpenRA.Launcher/Program.cs:line 3
```

**Comment by PunkPun** [comment]
this crash log is unhelpful. The replay is always necessary, and it it possible that this bug was already fixed

**Comment by anvilvapre** [comment]
[ra-2024-05-26T194414Z.orarep.gz](https://github.com/OpenRA/OpenRA/files/15448975/ra-2024-05-26T194414Z.orarep.gz)

**Comment by classCommandR** [comment]
Here is another occurrence of the same issue. In the screenshot you see a red blackhawk aircraft with infantry inside trying to land next to a harvester in the bottom left area. Then, the game immediately crashes.
![2024-05-29_09-41](https://github.com/OpenRA/OpenRA/assets/24455624/6c65f9f0-29b8-41cb-a7d7-66001e0616cc)
[ra-2024-05-29T065611Z.orarep.gz](https://github.com/OpenRA/OpenRA/files/15480800/ra-2024-05-29T065611Z.orarep.gz)
`[user1@fedora Logs]$ cat exception-2024-05-29T071007Z.log 
OpenRA engine version release-20231010
OpenRA Language: en
Red Alert mod version release-20231010
on map bb885c240b925f5870c7d586d3fe45f2059fde9b (Ensio Kaivo by Jopani Kansaro).
Date: 2024-05-29 07:10:07Z
Operating System: Linux (X64, Unix 6.8.9.100)
Runtime Version: .NET CLR 6.0.23
Installed Language: de (Installed) de (Current) de (Current UI)
Exception of type 'System.InvalidOperationException': Cannot AddInfluence until previous influence is removed with RemoveInfluence
   at OpenRA.Mods.Common.Traits.Aircraft.AddInfluence(ValueTuple'2[] landingCells) in /root/parts/openra/build/OpenRA.Mods.Common/Traits/Air/Aircraft.cs:line 870
   at OpenRA.Mods.Common.Traits.Aircraft.AddInfluence(CPos landingCell) in /root/parts/openra/build/OpenRA.Mods.Common/Traits/Air/Aircraft.cs:line 880
   at OpenRA.Mods.Common.Activities.Land.Tick(Actor self) in /root/parts/openra/build/OpenRA.Mods.Common/Activities/Air/Land.cs:line 232
   at OpenRA.Activities.Activity.TickOuter(Actor self) in /root/parts/openra/build/OpenRA.Game/Activities/Activity.cs:line 108
   at OpenRA.Traits.ActivityUtils.RunActivity(Actor self, Activity act) in /root/parts/openra/build/OpenRA.Game/Traits/ActivityUtils.cs:line 29
   at OpenRA.Actor.Tick() in /root/parts/openra/build/OpenRA.Game/Actor.cs:line 262
   at OpenRA.World.Tick() in /root/parts/openra/build/OpenRA.Game/World.cs:line 446
   at OpenRA.Game.InnerLogicTick(OrderManager orderManager) in /root/parts/openra/build/OpenRA.Game/Game.cs:line 631
   at OpenRA.Game.LogicTick() in /root/parts/openra/build/OpenRA.Game/Game.cs:line 646
   at OpenRA.Game.Loop() in /root/parts/openra/build/OpenRA.Game/Game.cs:line 815
   at OpenRA.Game.Run() in /root/parts/openra/build/OpenRA.Game/Game.cs:line 868
   at OpenRA.Game.InitializeAndRun(String[] args) in /root/parts/openra/build/OpenRA.Game/Game.cs:line 306
   at OpenRA.Launcher.Program.Main(String[] args) in /root/parts/openra/build/OpenRA.Launcher/Program.cs:line 32
`

**Comment by classCommandR** [comment]
Related issue: https://github.com/OpenRA/OpenRA/pull/21346
@PunkPun I have recreated the crash myself in Skirmish now, see description in the screenshot attached.
Correction: the force-land waypoints command is CTRL+ALT+SHIFT+CLICK on the areas
![2024-05-29_11-14](https://github.com/OpenRA/OpenRA/assets/24455624/ffd1c02d-aa8b-492e-a296-895cfcf154b1)

**Comment by eengmann** [comment]
![Uploading 2024-05-29_11-14.png…]()

**Comment by anvilvapre** [comment]
Proposal to improve the exception message to include the type of both influences and the actor it relates to etc.

**Comment by anvilvapre** [comment]
A better exception message does not seem needed as this message always relates to Aircraft. I thought Influences might be Traits or alike. But the Aircraft is the influence.

Aircraft.cs
```
public void AddInfluence((CPos, SubCell)[] landingCells)                                                
                {                                                                                                       
                        if (HasInfluence())                                                                             
                                throw new InvalidOperationException(                                                    
                                        $"Cannot {nameof(AddInfluence)} until previous influence is removed with {nameof(RemoveInfluence)}");                                                                                                   
                                                                                                                        
                        this.landingCells = landingCells;                                                               
                        if (self.IsInWorld)                                                                             
                                self.World.ActorMap.AddInfluence(self, this);                                           
                }         
```

Triggered from Land.Tick. Which aligns with the above message on how to reproduce.
```
aircraft.AddInfluence(landingCell);
```
In the Air activity sub dir this seems to be the only location where AddInfluence is called.

Aircraft.HasInfluence:
```                
public bool HasInfluence()                                                                             
                {                                                                                                      
                        return landingCells.Length > 0;                                                                
                }

```

And so it suggests that the cause is that `landingCells` is set multiple times.

`landingCells` is only set from `AddInfluence` and from `SetPosition`. The last is public and is also called internally from different locations.

Perhaps this gives a lead. 

Perhaps multiple occurences of a `Land` activity?

**Comment by PunkPun** [comment]
the issue is always that it was forgotten to remove influence. This crash was added to prevent bogus occupied cells

**Comment by anvilvapre** [comment]
Based on replay of enggman.

Sequence of events for actor that causes crash. Only last three are relevant. 
Logging when SetPosition is called. Also Logging when the line is reached within SetPosition where landingCells is assigned. And logging when AddInfluence is called.


SetPosition is called where landingCells is assigned. Followed by AddInfluence.

So likely `VerticalTakeOffOrLandTick`.

```
==== Aircraft.SetPosition $mh60 611, $   at OpenRA.Mods.Common.Traits.Aircraft.SetPosition(Actor self, WPos pos)
   at OpenRA.Mods.Common.Activities.Fly.VerticalTakeOffOrLandTick(Actor self, Aircraft aircraft, WAngle desiredFacing, WDist desiredAltitude, Boolean idleTurn)
   at OpenRA.Mods.Common.Activities.Land.Tick(Actor self)
   at OpenRA.Activities.Activity.TickOuter(Actor self)
   at OpenRA.Traits.ActivityUtils.RunActivity(Actor self, Activity act)
   at OpenRA.Actor.Tick()
   at OpenRA.World.Tick()
   at OpenRA.Game.InnerLogicTick(OrderManager orderManager)
   at OpenRA.Game.LogicTick()
   at OpenRA.Game.Loop()
   at OpenRA.Game.Run()
   at OpenRA.Game.InitializeAndRun(String[] args)
   at OpenRA.Launcher.Program.Main(String[] args)

==== Aircraft.SetPosition $mh60 611, $   at OpenRA.Mods.Common.Traits.Aircraft.SetPosition(Actor self, WPos pos)
   at OpenRA.Mods.Common.Activities.Fly.VerticalTakeOffOrLandTick(Actor self, Aircraft aircraft, WAngle desiredFacing, WDist desiredAltitude, Boolean idleTurn)
   at OpenRA.Mods.Common.Activities.Land.Tick(Actor self)
   at OpenRA.Activities.Activity.TickOuter(Actor self)
   at OpenRA.Traits.ActivityUtils.RunActivity(Actor self, Activity act)
   at OpenRA.Actor.Tick()
   at OpenRA.World.Tick()
   at OpenRA.Game.InnerLogicTick(OrderManager orderManager)
   at OpenRA.Game.LogicTick()
   at OpenRA.Game.Loop()
   at OpenRA.Game.Run()
   at OpenRA.Game.InitializeAndRun(String[] args)
   at OpenRA.Launcher.Program.Main(String[] args)

==== Aircraft.SetPosition assignLandingCells $mh60 611, $   at OpenRA.Mods.Common.Traits.Aircraft.SetPosition(Actor self, WPos pos)
   at OpenRA.Mods.Common.Activities.Fly.VerticalTakeOffOrLandTick(Actor self, Aircraft aircraft, WAngle desiredFacing, WDist desiredAltitude, Boolean idleTurn)
   at OpenRA.Mods.Common.Activities.Land.Tick(Actor self)
   at OpenRA.Activities.Activity.TickOuter(Actor self)
   at OpenRA.Traits.ActivityUtils.RunActivity(Actor self, Activity act)
   at OpenRA.Actor.Tick()
   at OpenRA.World.Tick()
   at OpenRA.Game.InnerLogicTick(OrderManager orderManager)
   at OpenRA.Game.LogicTick()
   at OpenRA.Game.Loop()
   at OpenRA.Game.Run()
   at OpenRA.Game.InitializeAndRun(String[] args)
   at OpenRA.Launcher.Program.Main(String[] args)

==== Aircraft.AddInfluence $mh60 611, $   at OpenRA.Mods.Common.Traits.Aircraft.AddInfluence(ValueTuple`2[] landingCells)
   at OpenRA.Mods.Common.Traits.Aircraft.AddInfluence(CPos landingCell)
   at OpenRA.Mods.Common.Activities.Land.Tick(Actor self)
   at OpenRA.Activities.Activity.TickOuter(Actor self)
   at OpenRA.Traits.ActivityUtils.RunActivity(Actor self, Activity act)
   at OpenRA.Actor.Tick()
   at OpenRA.World.Tick()
   at OpenRA.Game.InnerLogicTick(OrderManager orderManager)
   at OpenRA.Game.LogicTick()
   at OpenRA.Game.Loop()
   at OpenRA.Game.Run()
   at OpenRA.Game.InitializeAndRun(String[] args)
   at OpenRA.Launcher.Program.Main(String[] args)

OpenRA has encountered a fatal error.
```

**Comment by anvilvapre** [comment]
Since `landingCells` is only assigned from `SetPosition` if `landingCells` was already assigned prior - there should have been a call to `AddInfluence` prior.  

So if we look at all calls this is the full replay sequence for the actor. Ommiting the many calls to SetPosition where `landingCells` is not assigned.
```
==== Aircraft.AddInfluence $mh60 611, $   at OpenRA.Mods.Common.Traits.Aircraft.AddInfluence(ValueTuple`2[] landingCells)
==== Aircraft.SetPosition assignLandingCells $mh60 611, $   at OpenRA.Mods.Common.Traits.Aircraft.SetPosition(Actor self, WPos pos)
==== Aircraft.AddInfluence $mh60 611, $   at OpenRA.Mods.Common.Traits.Aircraft.AddInfluence(ValueTuple`2[] landingCells)
==== Aircraft.SetPosition assignLandingCells $mh60 611, $   at OpenRA.Mods.Common.Traits.Aircraft.SetPosition(Actor self, WPos pos)
==== Aircraft.AddInfluence $mh60 611, $   at OpenRA.Mods.Common.Traits.Aircraft.AddInfluence(ValueTuple`2[] landingCells)
==== Aircraft.SetPosition assignLandingCells $mh60 611, $   at OpenRA.Mods.Common.Traits.Aircraft.SetPosition(Actor self, WPos pos)
==== Aircraft.AddInfluence $mh60 611, $   at OpenRA.Mods.Common.Traits.Aircraft.AddInfluence(ValueTuple`2[] landingCells)
```

**Comment by RoosterDragon** [comment]
This has the same cause as #21302, which was fixed by #21346. We can close this.

If you run the replay with the fix applied, it no longer crashes and instead ends cleanly.
